### ISSUE TITLE
[data grid] Expandable rows are not accessible

### ISSUE THREAD

**Issue by ArmandRedgate** [mui/mui-x#10032]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example:

https://mui.com/x/react-data-grid/tree-data/

Steps:

1. Expand one of the rows
2. Inspect the accessibility tree

Order ID: 65425

### Current behavior 😯

The current data grid example for expandable rows is not accessible as it does not follow an the correct pattern.

To illustrate the issue, lets look at the component and its the accessibility tree:

![Component](https://github.com/mui/material-ui/assets/93141585/bdf6339a-9bdb-41c9-8e2c-c2b3140f5f15)

![Accessibility tree, annotated to show missing aria properties](https://github.com/mui/material-ui/assets/93141585/59a383ea-03e7-42b7-92a5-265f9f930353)

As the accessibility tree illustrates, there is nothing to relate to non-sighted users that one of the rows is expanded and that some of the rows are child rows of another.

### Expected behavior 🤔

When expandable rows are present the [Treegrid pattern](https://www.w3.org/WAI/ARIA/apg/patterns/treegrid/) should be employed.

From an aria property point of view, this means adding properties like:
- `role=treegrid` to indicate this is a tree grid
- `aria-level` to determine the depth
- `aria-expanded` to say if the current row is expanded (leave it `undefined` to say this is a leaf node)
- `aria-setsize` to determine how many children there are
- `aria-posinset` to say what item you are on at the current level

I've not had a proper look at the keyboard interactions. They may be fine as or or require some changes too.

### Context 🔦

We've started using MUI recently, and are looking to replace our expandable row patterns with poor accessibility with standard approaches, and were hoping MUI had an out-of-the-box solution.

We pay for MUI-X, and were hoping data-grid would implement these correctly, but that doesn't seem to currently be the case.

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```

Chrome Version 115.0.5790.171 (Official Build) (64-bit)

</details>

**Comment by arminmeh** [comment]
@ArmandRedgate just wanted to let you know that I checked the keyboard interaction and found that it is also not fully aligned with the specs, so I have made another issue to address that
https://github.com/mui/mui-x/issues/14204

**Comment by ArmandRedgate** [comment]
@arminmeh thank you very much 🙂
